File: server/fakes/focus_inspector.go
// screenreader-mcp fakes -- FakeFocusInspector: the FocusInspector port double.
// Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.
//
// ROLE: test double for domain/ports/focus_inspector.go.
// USED BY: the get_focus_info tool controller tests.
package fakes

import (
	"sync"

	"github.com/marlon-sousa/screen-readers-mcp/server/domain/ports"
)

type FakeFocusInspector struct {
	mu    sync.Mutex
	focus ports.FocusInfo
	err   error
}

var _ ports.FocusInspector = (*FakeFocusInspector)(nil)

func NewFakeFocusInspector() *FakeFocusInspector { return &FakeFocusInspector{} }

func (f *FakeFocusInspector) SetFocus(focus ports.FocusInfo) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.focus = focus
}

func (f *FakeFocusInspector) FailWith(err error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	f.err = err
}

func (f *FakeFocusInspector) FocusInfo() (ports.FocusInfo, error) {
	f.mu.Lock()
	defer f.mu.Unlock()
	if f.err != nil {
		return ports.FocusInfo{}, f.err
	}
	return f.focus, nil
}

File: bridges/nvda/tests/fakes/focus_inspector.py
# nvdaMcpBridge tests -- FakeFocusInspector, standing in for the FocusInspector port.
# Copyright (C) 2026 Marlon Brandao de Sousa. GPL-2. See COPYING.txt.

from __future__ import annotations

from nvdaMcpBridge.domain.ports.focus_inspector import FocusInfo, FocusInspector


class FakeFocusInspector(FocusInspector):
	def __init__(self) -> None:
		self.calls: list[None] = []
		self.info: FocusInfo = FocusInfo(
			name="Test Button",
			role="BUTTON",
			states=["FOCUSABLE", "FOCUSED"],
			value=None,
			app_module="test_app",
		)

	def focus_info(self) -> FocusInfo:
		self.calls.append(None)
		return self.info
